examples/rsa-example.cpp

```diff
@@ -0,0 +1,95 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2020 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+#include <iostream>
+#include <mococrw/asymmetric_crypto_ctx.h>
+#include <mococrw/padding_mode.h>
+
+using namespace mococrw;
+
+std::vector<uint8_t> encrypt_rsa(const std::shared_ptr<AsymmetricPublicKey> &pubKey,
+                                 const std::shared_ptr<RSAEncryptionPadding> &padding,
+                                 const std::vector<uint8_t> &data)
+{
+    RSAEncryptionPublicKeyCtx rsaPubCtx(*pubKey, padding);
+
+    std::vector<uint8_t> ciphertext;
+    try {
+        ciphertext = rsaPubCtx.encrypt(data);
+    } catch (const mococrw::MoCOCrWException &e) {
+        /* Possible reasons:
+         * - Error in openssl
+         * - Message size is not suited for encryption
+         * - Message to big
+         * - Encryption context is not set
+         */
+        std::cerr << "Failure in crypto engine: ";
+        std::cerr << e.what() << std::endl;
+        exit(EXIT_FAILURE);
+    }
+
+    return ciphertext;
+}
+
+std::vector<uint8_t> decrypt_rsa(const std::shared_ptr<AsymmetricPrivateKey> &privKey,
+                                 const std::shared_ptr<RSAEncryptionPadding> &padding,
+                                 const std::vector<uint8_t> &ciphertext) {
+    std::vector<uint8_t> plaintext;
+    /* Padding is optional. Default is OAEP with SHA256 and MGF1(SHA256) as mask generation function */
+    RSAEncryptionPrivateKeyCtx rsaPrivCtx(*privKey, padding);
+    try {
+        plaintext = rsaPrivCtx.decrypt(ciphertext);
+    } catch (const MoCOCrWException &e) {
+        /* Possible reasons:
+         * - Error in openssl (padding, decryption, ...)
+         */
+        std::cerr << "Error decrypting data." << std::endl;
+        std::cerr << e.what();
+        exit(EXIT_FAILURE);
+    }
+
+   return plaintext;
+}
+
+int main(void)
+{
+    auto keyPair = std::make_shared<AsymmetricKeypair>(AsymmetricKeypair::generateRSA());
```

Comment:
An `AsymmetricKeypair` is a wrapper around a `std::shared_ptr` and thus lightweight. We should copy it directly rather than creating shared pointers to it. This also applies to the `encrypt_rsa` and `decrypt_rsa` interfaces above.